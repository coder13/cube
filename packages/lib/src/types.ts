export interface PieceSetWithOrientation {
  perm: number[];
  orient: number[];
}

export interface Puzzle {
  corners: PieceSetWithOrientation;
  edges: PieceSetWithOrientation;
  centers: number[];
}

export interface MoveDef {
  corners?: PieceSetWithOrientation;
  edges?: PieceSetWithOrientation;
  centers?: number[];
}

export type Face = "up" | "down" | "left" | "right" | "back" | "front";

export enum Center {
  up = 0,
  left = 1,
  front = 2,
  right = 3,
  back = 4,
  down = 5,
}
