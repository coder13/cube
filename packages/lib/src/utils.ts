import { MoveDef, Puzzle } from "./types";

export const swap = (arr: number[], i: number, j: number) => {
  let newArr = [...arr];
  [newArr[i], newArr[j]] = [newArr[j], newArr[i]];
  return newArr;
};

export type PermutationFunc = (
  how: number[]
) => (p: number, i: number, what: number[]) => number;
export type OrientFunc = (
  how: number[],
  n: number
) => (p: number, i: number) => number;

export const perm: PermutationFunc = (how) => (_, i, what) => what[how[i]];

export const orient: OrientFunc = (how, n) => (p, i) => (p + how[i]) % n;

export const solved = (): Puzzle => ({
  corners: {
    perm: [0, 1, 2, 3, 4, 5, 6, 7],
    orient: [0, 0, 0, 0, 0, 0, 0, 0],
  },
  edges: {
    perm: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11],
    orient: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  },
  centers: [0, 1, 2, 3, 4, 5],
});

export const cycle = (_move: MoveDef, _cube: MoveDef) => {
  const cube: Puzzle = { ...solved(), ..._cube };
  const move: Puzzle = { ...solved(), ..._move }; // fill in the gaps

  const cp = perm(move.corners.perm);
  const co = orient(move.corners.orient, 3);
  const ep = perm(move.edges.perm);
  const eo = orient(move.edges.orient, 2);

  return {
    corners: {
      perm: cube.corners.perm.map(cp),
      orient: cube.corners.orient.map(co).map(cp),
    },
    edges: {
      perm: cube.edges.perm.map(ep),
      orient: cube.edges.orient.map(eo).map(ep),
    },
    centers: cube.centers.map(perm(move.centers)),
  };
};

export const combine = (moves: MoveDef[]): Puzzle => {
  return moves.reduce(cycle, solved()) as Puzzle;
};
