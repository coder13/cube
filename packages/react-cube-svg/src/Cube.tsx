import { Cube as CubeState, Face as FaceType } from "@cp/lib";
import Face from "./Face";

/* Colors: */
const Black = "#000000";
const White = "#FFFFFF";
const Yellow = "#FEFE00";
const Red = "#EE0000";
const Orange = "#FFA100";
const Blue = "#0000F2";
const Green = "#00D800";

export type ColorScheme = Record<FaceType, string>;

export interface CubeProps {
  scramble?: string;
  puzzle?: number;
  size?: number;
  lineWidth?: number;
  colorScheme?: ColorScheme;
  lineColor?: string;
  spacing?: number;
}

const Faces: Array<{
  face: FaceType;
  x: number;
  y: number;
}> = [
  {
    face: "up",
    x: 3,
    y: 0,
  },
  {
    face: "left",
    x: 0,
    y: 3,
  },
  {
    face: "front",
    x: 3,
    y: 3,
  },
  {
    face: "down",
    x: 3,
    y: 6,
  },
  {
    face: "right",
    x: 6,
    y: 3,
  },
  {
    face: "back",
    x: 9,
    y: 3,
  },
];

export const CubeRenderer = ({
  scramble = "",
  puzzle = 3,
  size = 200,
  colorScheme = {
    up: White,
    down: Yellow,
    front: Green,
    back: Blue,
    left: Orange,
    right: Red,
  },
  lineColor = Black,
  lineWidth = 1,
  spacing = 0.9,
  ...props
}: CubeProps) => {
  const state = new CubeState(scramble);

  const getFace = (f: FaceType) =>
    state.getFace(f).map((row) => row.map((s) => colorScheme[s]));

  return (
    <svg
      width={size}
      height={(size * 3) / 4}
      viewBox={`0 0 ${puzzle * 4} ${puzzle * 3}`}
      {...props}
    >
      {Faces.map(({ face, x, y }) => (
        <Face
          key={face}
          x={x}
          y={y}
          puzzle={puzzle}
          size={spacing}
          state={getFace(face)}
          lineColor={lineColor}
          lineWidth={lineWidth}
        />
      ))}
    </svg>
  );
};
