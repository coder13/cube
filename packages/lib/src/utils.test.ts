import Cube, { Moves } from "./cube";
import { PieceSetWithOrientation } from "./types";
import { combine, cycle, solved } from "./utils";
import { describe, expect, test } from "bun:test";

describe("cycle", () => {
  test("should return the correct cycle for corners and edges", () => {
    expect(cycle(solved(), solved())).toMatchObject(solved());

    expect(cycle(Moves["U"], solved()).corners).toMatchObject(
      Moves["U"].corners as PieceSetWithOrientation
    );
    expect(cycle(Moves["U"], solved()).edges).toMatchObject(
      Moves["U"].edges as PieceSetWithOrientation
    );
  });

  test("inverting moves", () => {
    expect(cycle(Moves["U"], cycle(Moves["U'"], solved()))).toMatchObject(
      solved()
    );

    const sexy = [Moves["R"], Moves["U"], Moves["R'"], Moves["U'"]];
    const combinedMoves = combine(sexy);

    expect(
      combine([
        combinedMoves,
        combinedMoves,
        combinedMoves,
        combinedMoves,
        combinedMoves,
        combinedMoves,
      ])
    ).toMatchObject(solved());
  });

  test("should combine moves correctly", () => {
    const cube = cycle(Moves["U"], solved());
    const cube2 = cycle(Moves["U'"], cube);
    expect(cube2).toMatchObject(solved());

    expect(combine([Moves["U"], Moves["U'"]])).toMatchObject(solved());

    const cube3 = new Cube("R U R' U R U2 R' U2");
    expect(cube3.getFace("up")).toMatchSnapshot();
    expect(cube3.getFace("down")).toMatchSnapshot();
  });
});
