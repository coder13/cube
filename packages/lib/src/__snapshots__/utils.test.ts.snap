// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`cycle should combine moves correctly 1`] = `
[
  [
    "back",
    "up",
    "right",
  ],
  [
    "up",
    "up",
    "up",
  ],
  [
    "up",
    "up",
    "front",
  ],
]
`;

exports[`cycle should combine moves correctly 2`] = `
[
  [
    "down",
    "down",
    "down",
  ],
  [
    "down",
    "down",
    "down",
  ],
  [
    "down",
    "down",
    "down",
  ],
]
`;

exports[`cycle should combine moves correctly 3`] = `
Cube {
  "centers": [
    0,
    1,
    2,
    3,
    4,
    5,
  ],
  "corners": {
    "orient": [
      2,
      1,
      1,
      2,
      0,
      1,
      2,
      0,
    ],
    "perm": [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
    ],
  },
  "edges": {
    "orient": [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    "perm": [
      2,
      0,
      1,
      3,
      4,
      5,
      6,
      7,
      8,
      9,
      10,
      11,
    ],
  },
  "moves": "",
}
`;
